File: 2349-design_a_number_container_system/ruby/main.rb
class NumberContainers
  def initialize
    @by_idx = {}
    @by_val = Hash.new { |hash, key| hash[key] = [] }
  end

  def change(index, number)
    if @by_idx.key?(index)
      old_number = @by_idx[index]
      return if old_number == number

      @by_val[old_number].delete(index)
      @by_val.delete(old_number) if @by_val[old_number].empty?
    end

    @by_idx[index] = number

    idx = @by_val[number].bsearch_index { |x| x >= index } || @by_val[number].size
    @by_val[number].insert(idx, index)
  end

  def find(number)
    @by_val[number].first || -1
  end
end

if __FILE__ == $0
  obj = NumberContainers.new

  if obj.find(10) != -1
    puts "1. Test failed, expected -1"
    exit
  end

  obj.change(2, 10)
  obj.change(1, 10)
  obj.change(3, 10)
  obj.change(5, 10)

  if obj.find(10) != 1
    puts "2. Test failed, expected 1"
    exit
  end

  obj.change(1, 20)

  if obj.find(10) != 2
    puts "3. Test failed, expected 2"
    exit
  end

  puts "Passed"
end

File: 2349-design_a_number_container_system/go/main.go
package main

import (
	"fmt"
	"slices"
)

type NumberContainers struct {
	byIdx map[int]int
	byVal map[int][]int
}

func Constructor() NumberContainers {
	return NumberContainers{
		byIdx: make(map[int]int),
		byVal: make(map[int][]int),
	}
}

func (this *NumberContainers) Change(index int, number int) {
	if val, ok := this.byIdx[index]; ok {
		if val == number {
			return
		}
		idx, _ := slices.BinarySearch(this.byVal[val], index)
		this.byVal[val] = append(this.byVal[val][:idx], this.byVal[val][idx+1:]...)
	}
	this.byIdx[index] = number
	if _, ok := this.byVal[number]; !ok {
		this.byVal[number] = []int{}
	}
	idx, _ := slices.BinarySearch(this.byVal[number], index)
	this.byVal[number] = append(this.byVal[number][:idx], append([]int{index}, this.byVal[number][idx:]...)...)
}

func (this *NumberContainers) Find(number int) int {
	if indexes, ok := this.byVal[number]; ok && len(indexes) > 0 {
		return indexes[0]
	}
	return -1
}

func main() {
	obj := Constructor()
	if val := obj.Find(10); val != -1 {
		fmt.Printf("1. Test failed, expected -1, got %d", val)
		return
	}
	obj.Change(2, 10)
	obj.Change(1, 10)
	obj.Change(3, 10)
	obj.Change(5, 10)
	if val := obj.Find(10); val != 1 {
		fmt.Printf("2. Test failed, expected 1, got %d", val)
		return
	}
	obj.Change(1, 20)
	if val := obj.Find(10); val != 2 {
		fmt.Printf("3. Test failed, expected 2, got %d", val)
		return
	}
	fmt.Println("Passed")
}
